### ISSUE TITLE
Error: ORA-01000: maximum open cursors exceeded --- With Connexion Pool

### ISSUE THREAD

**Issue by BlackMatrix971** [oracle/node-oracledb#106]
I created an asynchronous ExecuteQuery function that is able to execute query asynchronously so that if ever it's called 
inside a loop, the next query wait for the former one finishes before been executed.
The doCreatePool function is called once, at program initialization.
It creates a connexion pool and save it in  the Global variable config.Oracle_POOL so that the connexion pool remains alive, 
even after then doCreatePool returns.
'

Socket.io events call ExecuteQuery to execute queries (either one, or more in a loop).

Everything goes right but, after a certain amount of queries executed, 
I get an ORA-1000 error : "Error: ORA-01000: maximum open cursors exceeded"
either for select, update and insert queries.

Normally the Connexion pool is supposed to avoid this issue ?

What is wong in my code ?

```
doCreatePool (); //Create a connexion pool and save it in config.Oracle_POOL
//dbConfig={....}

function doCreatePool () {
    dbConfig=config.Settings.DataBaseConnexionString;
    oracle.createPool (
      {
        user          : dbConfig.user,
        password      : dbConfig.password,
        connectString : dbConfig.connectString,
        poolMax       : 44,
        poolMin       : 2,
        poolIncrement : 5,
        poolTimeout   : 4
      },
      function(err, pool)
      {
        if (err) {
          console.error("createPool() callback: " + err.message);
          return;
        } else {
            config.Oracle_POOL = pool;
        }
     });      
}

function ExecuteQuery(Query, LstParam, callBack) {

	config.Oracle_POOL.getConnection( function(err, connection) {
		if (err){ 
			console.log("Oracle connexion pool --> err --> ", err);
			if (connection)
			connection.release(
                function(err) {
                	if (err) {
                		console.log("(1) connection.release --> err--> ",err)
                	}
                }
            );  
            callBack(err, { rows : [] });  	
		} else {
			var OracleQueryOptions={outFormat: oracle.OBJECT, maxRows : 500, autoCommit : true};
			connection.execute(Query, LstParam, OracleQueryOptions , function(err, results) {
				if (!err) {
					console.log("* connection.execute  --> err--> ",err, "Query --> ", Query);
				}
				connection.release( function(err2) {	                    	
                    	if (err2) {
                    		console.log("(2) connection.release --> err-->",err)
                    	}
						callBack(err, results);
                    }
                );
			});
		}
	});		
			
		
}
```

**Comment by dmcghan** [comment]
@BlackMatrix971 I don't see any really obvious problems here problems here. Have you seen my post on creating a wrapper module? https://jsao.io/2015/03/making-a-wrapper-module-for-the-node-js-driver-for-oracle-database/ I'd be curious if you get the same issue with that.

Are you only executing simple SQL statements or PL/SQL too?

Something to note... The ExecuteQuery function, as far as I can tell, doesn't do anything to help with loop execution queuing. It seems to be more of an encapsulation of: going to the pool, getting a connection, executing a statement, and releasing the connection. I made a similar method in my post called simpleExecute. 

If you want to process queries sequentially in a loop, have a look at the async module: https://github.com/caolan/async It's one of the most popular modules in Node for good reason, it rocks!

Also, have you seen this post regarding threads in Node.js? http://www.future-processing.pl/blog/on-problems-with-threads-in-node-js/ You should fix the bug in your code prior using the technique in the post. Otherwise your problem will become more intermittent, but it will still exist.

**Comment by BlackMatrix971** [comment]
Hi @dmcghan

I execute only  simple SQL statements (Select, and Insert).
According to my tests The ExecuteQuery function works realy asynchronously in the loop.
When I console log, I can notice that the order of the statements in the loop is the same than the query execution.

In the statement options, instead of maxRows : 500, I overdrive maxRows : 100.

It's seems to work for now, I don't kown for how many times.
I let the App running, mean while I will try your advice and let you know the result.

**Comment by cjbj** [comment]
Have you tried different statement cache sizes?  Do you actually need to increase the database cursors?

**Comment by alexsantos** [comment]
Might be a bit off topic (my experience with this same error but in java):
this happened to me when I forgot to close the preparedStatement before
closing the connection borrowed from a connection pool. Closing the
connection doesn't automatically closes the cursors. This was on 11gR2.

**Comment by dmcghan** [comment]
@BlackMatrix971 Imagine you have a loop with console.log inside: 
for (var x = 0; x < 100; x += 1) {
	console.log(x);
}

Of course x would log to the console sequentially. But now let's say you do something asynchronous right after the console log:

function executeQuery(iteration) {
   var randomTime = Math.random() * 1000; // random timeout between 0 and 1 seconds

   setTimeout(function() {
      console.log('async done for ' + iteration);
   }, randomTime);
}

for (var x = 0; x < 100; x += 1) {
   console.log('about to fire async for ' + x);
   executeQuery(x);
}

If you run that you'll see the same sequential console logs followed by completely random completions of the asynchronous functions. This is what's happening in your loops. Not only are the queries not being executed sequentially (they are being executed in parallel) but the results are coming back at random times.

**Comment by cjbj** [comment]
One other comment is that you have copied the example pool configuration from examples/webapp.js.  Those values just examples.  You probably don't want to copy them verbatim.

**Comment by BlackMatrix971** [comment]
Hi @dmcghan !!!
you're right. I solve this problem with a self-calling recursion fonction. when ExecuteQuery() finishes running, il call back the loop function again for lets say 100 times.

It works very fine.

I know where my problem comes from. By adding these lines of code before calling the getConnexion() function

 // NUMBER OF CONNCETIONS OPEN
    console.log("B4getConnexion -> ORACLE: CONNX OPEN: " + config.Oracle_POOL.connectionsOpen);

    // NUMBER OF CONNEXTIONS IN USE
    console.log("B4getConnexion -> ORACLE: CONNX IN USE: " + config.Oracle_POOL.connectionsInUse);

and this  at the end of  the getConnexion() function call :

 // NUMBER OF CONNCETIONS OPEN
    console.log("AFTER/getConnexion -> ORACLE: CONNX OPEN: " + config.Oracle_POOL.connectionsOpen);

    // NUMBER OF CONNEXTIONS IN USE
    console.log("AFTER/getConnexion -> ORACLE: CONNX IN USE: " + config.Oracle_POOL.connectionsInUse);


I have notice that config.Oracle_POOL.connectionsOpen increses until it reaches poolMax (44). This is why I get erros like (this is my opinion for now) : 

ORA-24418: Cannot open further sessions
ORA-24418: Cannot open further sessions
ORA-01000: maximum open cursors exceeded 

How can I avoid config.Oracle_POOL.connectionsOpen to increase ?.
Please note that console.log inform that RelaseConnexion is successfully called after ExecuteQuery and then getConnexion. That all sequence are executed sequentially.

If I don't use close connexion (which mean by executing several statement in the loop) config.Oracle_POOL.connections  stay stable.
But at the end of the loop I clause the connexion, and when an event occur, it starts a new loop and I create a new connexion, then config.Oracle_POOL.connections increases...

**Comment by alexsantos** [comment]
The number of connections or sessions is application dependant. If you have
1000 simultaneous users on your application you have to configure your
database for such.
On the other hand, the open cursors error is, IMHO, a programming error
like I mentioned earlier.

Alex
Em 06/06/2015 11:11, "BlackMatrix971" <notifications@github.com> escreveu:

**Comment by dmcghan** [comment]
@BlackMatrix971 This would be a lot easier if you could either show us your code or create some mockup code that demonstrates your issue that you can share.

Since you are executing statements in series, why don't you just reuse the same connection? Get the connection once, then go into your recursive loop reusing the same connection many times, then release it after.

Maybe I should write a blog post on this and related topics...

**Comment by BlackMatrix971** [comment]
Ok.

@dmcghan 
I'll write a demo code as it a bit big ;-) I'll remove not need computation code with in the loop an post.
I reuse the same connexion in a session (one loop) all the statement are executed inside.

When the loop end, I release then connexion. ReleaseConnection() doesn't give error. 

When an other event occur I call getConnexion. but config.Oracle_POOL.connections increases.
A new loop starts ... and so one...

After some hours  but config.Oracle_POOL.connections reaches poolmax, and bang !!!

I write the code for you to understand better.

@alexsantos 
Thanks a lot. It's thank to you help I discover the origine of my problem.
I'm not far to find the key ;-)
I'm patient. 

Thanks to all.

I'll be back with the code.

**Comment by BlackMatrix971** [comment]
A ok.

Thanks. I never use prepareStatement.

I gona read about it. 


     Le Vendredi 5 juin 2015 18h49, Christopher Jones <notifications@github.com> a écrit :
   

 One other comment is that you have copied the example pool configuration from examples/webapp.js. Those values just examples. You probably don't want to copy them verbatim.—

**Comment by cjbj** [comment]
There are no explicit 'prepare statement' calls in node-oracledb.  The statement caching feature of the underlying Oracle libraries has the same effect.  There is a statement cache per connection-pool session (and for each non pooled connection). I will improve the doc on this.

A runnable example of what you are doing would be great.

**Comment by BlackMatrix971** [comment]
Hi All !!!

This is the runnable code :

***************************************************
SQL Scrupt :



CREATE TABLE Author
( AuthorKey varchar2(10) NOT NULL,
  Adresses varchar2(10) NOT NULL,
  NickName varchar2(50)
);
INSERT ALL
	INTO Author (AuthorKey, Adresses, NickName) VALUES ('A0', '2', 'SeanPAUL')
	INTO Author (AuthorKey, Adresses, NickName) VALUES ('A1', '5', 'K2000')
	INTO Author (AuthorKey, Adresses, NickName) VALUES ('A2', '1', 'PhilCollins')		
SELECT * FROM dual;


CREATE TABLE AuthADR
( AdresseKey varchar2(10) NOT NULL,
  Email varchar2(50) NOT NULL,
  Street varchar2(50),
  PostalCode number(10) NOT NULL,
  City varchar2(50) NOT NULL,
  Country varchar2(50)
);


INSERT ALL
	INTO AuthADR (AdresseKey, Email, Street, PostalCode, City, Country) VALUES ('1', 'black1@matrix.com', 'My Street1', 53142,'New York', 'USA')
	INTO AuthADR (AdresseKey, Email, Street, PostalCode, City, Country) VALUES ('2', 'black2@matrix.com', 'My Street2', 53142,'New York', 'France')
	INTO AuthADR (AdresseKey, Email, Street, PostalCode, City, Country) VALUES ('3', 'black3@matrix.com', 'My Street3', 53142,'New York', 'Espana')	
	INTO AuthADR (AdresseKey, Email, Street, PostalCode, City, Country) VALUES ('4', 'black4@matrix.com', 'My Street4', 53142,'New York', 'England')
	INTO AuthADR (AdresseKey, Email, Street, PostalCode, City, Country) VALUES ('5', 'black5@matrix.com', 'My Street5', 53142,'New York', 'Switzeland')
	INTO AuthADR (AdresseKey, Email, Street, PostalCode, City, Country) VALUES ('6', 'black6@matrix.com', 'My Street6', 53142,'New York', 'ITALY')		
SELECT * FROM dual;



CREATE TABLE Countries
( CountryKey varchar2(50) NOT NULL,
  XCoor number(10) NOT NULL,
  YCoor number(10),
  Land varchar2(50) NOT NULL
);
INSERT ALL
	INTO Countries (CountryKey, XCoor, YCoor, Land) VALUES ('USA', 125.341, 125.641, 'USA')
	INTO Countries (CountryKey, XCoor, YCoor, Land) VALUES ('USA', 125.342, 125.541, 'Porto rico')
	INTO Countries (CountryKey, XCoor, YCoor, Land) VALUES ('France', 125.343, 125.441, 'Réunion')	
	INTO Countries (CountryKey, XCoor, YCoor, Land) VALUES ('France', 125.344, 125.341, 'St-Pierre')
	INTO Countries (CountryKey, XCoor, YCoor, Land) VALUES ('France', 125.345, 125.241, 'Hexagone')
	INTO Countries (CountryKey, XCoor, YCoor, Land) VALUES ('France', 125.346, 125.141, 'Miquelon')	
SELECT * FROM dual;


***************************************************


Table Author
======================
AuthorKey 	|		Adresses 		| 	NickName	| 
=======================================================
	A0		|			2			|	SeanPAUL
-------------------------------------------------------
	A1		|			5 			|	K2000
-------------------------------------------------------
	A2		|			1			|	PhilCollins
-------------------------------------------------------



---------------------
	Table Adresses	
===================================================================================================
	AdresseKey 	| 		Email			| 	Street 		| 	PostalCode	|	City		| 	Country
===================================================================================================
		1 		| 	black1@matrix.com	| My Street1	| 	53142		|	New York	| 	USA
---------------------------------------------------------------------------------------------------
		2 		| 	black2@matrix.com	| My Street2	| 	75015		|	  Paris		| 	France		
---------------------------------------------------------------------------------------------------
		3 		| 	black3@matrix.com	| My Street3	| 	35014		|	Barcelona	| 	Espana
---------------------------------------------------------------------------------------------------
		4 		| 	black4@matrix.com	| My Street4	| 	15027		|	 LONDON		| 	England		
---------------------------------------------------------------------------------------------------
		5 		| 	black5@matrix.com	| My Street5	| 	53142		|	  Bern		| 	Switzeland
---------------------------------------------------------------------------------------------------
		6 		| 	black6@matrix.com	| My Street6	| 	53142		|	  Roma		| 	ITALY						
---------------------------------------------------------------------------------------------------




---------------------
	Table Countries	
=============================================================
	CountryKey 	| 	XCoor	| 	YCoor 	| 	Land		|	
=============================================================
		USA 	| 	125.341	| 125.645	| 		USA		|	
-------------------------------------------------------------
		USA 	| 	125.342	| 125.545	| 	Porto rico	|
-------------------------------------------------------------
		France 	| 	125.343	| 125.445	| 	Réunion		|	
-------------------------------------------------------------
		France 	| 	125.344	| 125.345	| 	St-Pierre	|	 
-------------------------------------------------------------
		France 	| 	125.345	| 125.245	| 	Hexagone	|	 
-------------------------------------------------------------
		France 	| 	125.346	| 125.145	| 	Miquelon	|	  					
-------------------------------------------------------------

***************************************************
Javascript
***************************************************

var oracle = require('oracledb');

/*
=============================
==		DataStructure		=
=============================
*/
var TblStruc= {
	primaryKey : "AUTHORKEY",
	table : "Author",
	feildList : [
		{
			feildName : "AUTHORKEY",
			subFeilds : {}
		},	
		{
			feildName : "ADRESSES",
			subFeilds : {
				primaryKey : "ADRESSEKEY",
				table : "AuthADR",
				feildList : [
					{
						feildName : "ADRESSEKEY",
						subFeilds : {}
					},	
					{
						feildName : "EMAIL",
						subFeilds : {}
					},	
					{
						feildName : "COUNTRY",
						subFeilds : {
							primaryKey : "COUNTRYKEY",
							table : "COUNTRIES",
							feildList : [
								{
									feildName : "XCOOR",
									subFeilds : {}
								},	
								{
									feildName : "YCOOR",
									subFeilds : {}
								},	
								{
									feildName : "LAND",
									subFeilds : {}
								},											
							]
						}
					},											
				]
			}
		},
		{
			feildName : "NICKNAME",
			subFeilds : {}
		},				
	]

}

/*
=============================
==		Config Variable		=
=============================
*/

var config={
	Oracle_POOL : {},
	Connexions : {},
	stmtCounter : {},
	Settings : {
		DataBaseConnexionString : {
            "connectString" : "(DESCRIPTION =(ADDRESS_LIST =(ADDRESS = (PROTOCOL = TCP)(HOST = 10.42.0.76)(PORT = 1521)))(CONNECT_DATA =(SERVICE_NAME = ORCL)))",
            "user" : "matrix",
            "password" : "black"			
		}
	}	
};
/*
=========================
==		APP Code		=
=========================
*/
console.log("initializing...")
doCreatePool (function(){
	getConnexion(function(){
		console.log("doCreatePool -> getConnexion");

		setInterval(function(){
		 	MakeDOC({PRIMARYKEY : "A0"}, TblStruc, function(err, data) {
		 		if (err) {
		 			console.log("MakeDOC -> err", err, data)
		 		}
		 	}); 
			 	//Simulate event :
			 	//	socket.on('MakeDOC', function(Response){		
				//		MakeDOC(Response, TblStruc);
				//	})
		}, 3000);

	})
})


function MakeDOC(data, TblStruc, MainCallBack) {
	var newOBJ={}, LoopStack=[];

	MakeQuery(TblStruc, data.PRIMARYKEY, function(errX, Statement){
		if (!errX) {
			ExecuteQuery(Statement, [], function (err, results) {
				if (err || (results && results.rows && results.rows.length==0)) {
					MainCallBack(err, Statement);
				} else {
					AsyncForEach = function(Items, Index) { 
						Add2DOC(Items[Index], newOBJ, results.rows, function (subOBJ){
							if (Items.length-1>Index) { 
								AsyncForEach(Items, ++Index); 
							} else { 
								//New JSON Object Building Complete:
								console.log("New JSON Object Building Complete : ", JSON.stringify(newOBJ, null, 2) );
							}
						}, true);
					}
					AsyncForEach(TblStruc.feildList, 0);
				}
			});
		} 
	})

	function Add2DOC (item, subOBJ, results2, callback, Last){
		var temp = item.subFeilds.feildList;
		if (! subOBJ[item.feildName]) subOBJ[item.feildName]=[];
		AsyncForEach2 = function(Items2, Index2) { //Result rows Iteration

			function checkAsyncForEach2Loop(callback, subOBJ) { //Call AsyncForEach2 loop till the end of Result rows
				//console.log("LoopStack", LoopStack, "subOBJ : ", subOBJ, callback)
				if (LoopStack.length>0) {
					LoopStack.some(function (x, i) { 
						topLevel=LoopStack.length-1;
						PreviousLoop_OBJ=LoopStack[topLevel];
						if (PreviousLoop_OBJ.Items.length-1>PreviousLoop_OBJ.Index) {	
							AsyncForEach2(PreviousLoop_OBJ.Items, ++LoopStack[i].Index); 
							return true;
						} else {
							LoopStack.pop();
							if (LoopStack.length==0) {
								callback(subOBJ);	//End of loop return !!!
							}
						}
					});	
				} else {
					callback(subOBJ); //End of loop return !!!
				}			
			}	

			result=Items2[Index2]; 
		 	subOBJ[item.feildName].push(result[item.feildName]); 

			if(temp instanceof Array) { 
				MakeQuery(item, result[item.feildName], function(err, Statement){
					if (!err) {
						ExecuteQuery(Statement, [], function (err3, results3) {
							if (err3 || results3.rows.length==0) {
								checkAsyncForEach2Loop(callback, subOBJ);
							} else {
								LoopStack.push({"Items" : Items2, "Index" : Index2, "feildName" : item.feildName})
								subOBJ[item.feildName].push({});
								topLevel=subOBJ[item.feildName].length-1;
								var a=subOBJ[item.feildName][topLevel];
								AsyncForEach3 = function(Items3, Index3) {
									Add2DOC(Items3[Index3], a, results3.rows, function () {
										if (Items3.length-1>Index3) {
											AsyncForEach3(Items3, ++Index3);
										} else {
											checkAsyncForEach2Loop(callback, subOBJ);
										}	
									}, Items3.length-1>=Index3);
								}	
								AsyncForEach3(temp, 0);
							}
						});
					} else {
						checkAsyncForEach2Loop(callback, subOBJ);
					}
				});		
			} else {
				checkAsyncForEach2Loop(callback, subOBJ);
			}	 
		}
		AsyncForEach2(results2,0);
	}  
}

function MakeQuery(root, pk ,callback) {
	if ( pk ) {//
		var coma=Statement=""
		if (root.feildList) var feildList=root.feildList, table=root.table, primaryKey=root.primaryKey
		else var feildList=root.subFeilds.feildList, table=root.subFeilds.table, primaryKey=root.subFeilds.primaryKey;

		feildList.some(function(champ, index){
	        Statement=Statement+coma+champ.feildName+" "+champ.feildName;
	        coma=", ";
	        if (index==feildList.length-1) { 
	        	Statement="select "+primaryKey+" PRIMARYKEY, "+Statement+" FROM "+table+" WHERE "+primaryKey+" = '"+pk+"'";
				callback(false, Statement);
		    }        	
		});	
	} else {
		callback(true, pk); //Error when primaryKey value is null
	}
}


function getConnexion(callBack) {
	config.Oracle_POOL.getConnection( function(err, connection) {
		if (err){ 
			console.log("getConnection err --> ", err);
			if (connection)
			connection.release(
                function(err2) {
                	if (err2) {
                		console.log("getConnection -->  connection.release  --> err",err)
                	}
                }
            );  
            if (typeof callBack=="function") callBack(err, null);  	
		} else {

		 	// NUMBER OF CONNCETIONS OPEN
		    console.log("AFTER/getConnexion -> ORACLE: CONNX OPEN: " + config.Oracle_POOL.connectionsOpen);

		    // NUMBER OF CONNEXTIONS IN USE
		    console.log("AFTER/getConnexion -> ORACLE: CONNX IN USE: " + config.Oracle_POOL.connectionsInUse);

		    console.log("poolMax", config.Oracle_POOL.poolMax,  oracle.poolMax)

			config.Connexions=connection;
			if (typeof callBack=="function") callBack(err, connection); 
		}	
		
	});	
}	

function releaseConnection (callBack) {
	//Committing pending statement before releasing...
	config.Connexions.execute("select 1 from dual", [], {autoCommit : true} , function(err, results) {
		if (config.stmtCounter) config.stmtCounter.endAt = new Date();
		console.log("AutoComit at releaseConnection ** config.stmtCounter : ", config.stmtCounter);
		
		config.Connexions.release(function(err2) {
            	if (err2) {
            		console.log("config.Connexions.release --> err",err2);
            	} else console.log("Connexion released");
            	if (typeof callBack=="function") callBack(err, "Commit failed");  
        });  
	})	
}	


function ExecuteQuery(Query, LstParam, callBack) {
	if (config.Connexions){ 
		var OracleConfig={outFormat: oracle.OBJECT, maxRows : 200};
		
		if (!config.stmtCounter) config.stmtCounter={ "nonCommitable" : 0, Commitable : 0, startAt : new Date()};
		(/(^[ ]*Select)/gi).test(Query)?config.stmtCounter.nonCommitable++:config.stmtCounter.Commitable++;
		
		//Comming every 100 insert and update statement.
		if (config.stmtCounter.Commitable>100) {
			config.stmtCounter.Commitable=0;
			OracleConfig.autoCommit = true;
		}
		config.Connexions.execute(Query, LstParam, OracleConfig , function(err, results) {

	    	if (err && (/(ORA-01000)/gi).test(" "+err+" ")) 
	    		console.log("ORA-01000 error on stmtCounter : ", config.stmtCounter, " Query : ", Query);

			if (typeof callBack=="function") callBack(err, results);
			//setTimeout(function() {
			//	exit(0);
			//}, 60000)
		});	
	} else {
		if (typeof callBack=="function") callBack(true, results, "no connexion in config.Connexions "); 	
	}

}

function doCreatePool (callback) {
    dbConfig=config.Settings.DataBaseConnexionString;
    oracle.createPool (
      {
        user          : dbConfig.user,
        password      : dbConfig.password,
        connectString : dbConfig.connectString,
        poolMax       : 44,
        poolMin       : 2,
        poolIncrement : 5,
        poolTimeout   : 2,
        stmtCacheSize : 200
      },
      function(err, pool)
      {
        if (err) {
          console.error("createPool() callback: " + err.message);
          return;
        } else {
            config.Oracle_POOL = pool;
        }
        if (typeof callback=="function") callback(err, pool);
     });      
}

**Comment by BlackMatrix971** [comment]
OUTPUT result : 

New JSON Object Building Complete :  {
  "AUTHORKEY": [
    "A0"
  ],
  "ADRESSES": [
    "2",
    {
      "ADRESSEKEY": [
        "2"
      ],
      "EMAIL": [
        "black2@matrix.com"
      ],
      "COUNTRY": [
        "France",
        {
          "XCOOR": [
            125
          ],
          "YCOOR": [
            125
          ],
          "LAND": [
            "St-Pierre"
          ]
        },
        {
          "XCOOR": [
            125
          ],
          "YCOOR": [
            125
          ],
          "LAND": [
            "Hexagone"
          ]
        },
        {
          "XCOOR": [
            125
          ],
          "YCOOR": [
            125
          ],
          "LAND": [
            "St-Pierre"
          ]
        },
        {
          "XCOOR": [
            125
          ],
          "YCOOR": [
            125
          ],
          "LAND": [
            "Réunion"
          ]
        }
      ]
    }
  ],
  "NICKNAME": [
    "SeanPAUL"
  ]
}

**Comment by BlackMatrix971** [comment]
***************************************************
Javascript
***************************************************

				var oracle = require('oracledb');

				/*
				=============================
				==		DataStructure		=
				=============================
				*/
				var TblStruc= {
					primaryKey : "AUTHORKEY",
					table : "Author",
					feildList : [
						{
							feildName : "AUTHORKEY",
							subFeilds : {}
						},	
						{
							feildName : "ADRESSES",
							subFeilds : {
								primaryKey : "ADRESSEKEY",
								table : "AuthADR",
								feildList : [
									{
										feildName : "ADRESSEKEY",
										subFeilds : {}
									},	
									{
										feildName : "EMAIL",
										subFeilds : {}
									},	
									{
										feildName : "COUNTRY",
										subFeilds : {
											primaryKey : "COUNTRYKEY",
											table : "COUNTRIES",
											feildList : [
												{
													feildName : "XCOOR",
													subFeilds : {}
												},	
												{
													feildName : "YCOOR",
													subFeilds : {}
												},	
												{
													feildName : "LAND",
													subFeilds : {}
												},											
											]
										}
									},											
								]
							}
						},
						{
							feildName : "NICKNAME",
							subFeilds : {}
						},				
					]

				}

				/*
				=============================
				==		Config Variable		=
				=============================
				*/

				var config={
					Oracle_POOL : {},
					Connexions : {},
					stmtCounter : {},
					Settings : {
						DataBaseConnexionString : {
				            "connectString" : "(DESCRIPTION =(ADDRESS_LIST =(ADDRESS = (PROTOCOL = TCP)(HOST = 10.42.0.76)(PORT = 1521)))(CONNECT_DATA =(SERVICE_NAME = ORCL)))",
				            "user" : "matrix",
				            "password" : "black"			
						}
					}	
				};
				/*
				=========================
				==		APP Code		=
				=========================
				*/
				console.log("initializing...")
				doCreatePool (function(){
					getConnexion(function(){
						console.log("doCreatePool -> getConnexion");

						setInterval(function(){
						 	MakeDOC({PRIMARYKEY : "A0"}, TblStruc, function(err, data) {
						 		if (err) {
						 			console.log("MakeDOC -> err", err, data)
						 		}
						 	}); 
							 	//Simulate event :
							 	//	socket.on('MakeDOC', function(Response){		
								//		MakeDOC(Response, TblStruc);
								//	})
						}, 3000);

					})
				})


				function MakeDOC(data, TblStruc, MainCallBack) {
					var newOBJ={}, LoopStack=[];

					MakeQuery(TblStruc, data.PRIMARYKEY, function(errX, Statement){
						if (!errX) {
							ExecuteQuery(Statement, [], function (err, results) {
								if (err || (results && results.rows && results.rows.length==0)) {
									MainCallBack(err, Statement);
								} else {
									AsyncForEach = function(Items, Index) { 
										Add2DOC(Items[Index], newOBJ, results.rows, function (subOBJ){
											if (Items.length-1>Index) { 
												AsyncForEach(Items, ++Index); 
											} else { 
												//New JSON Object Building Complete:
												console.log("New JSON Object Building Complete : ", JSON.stringify(newOBJ, null, 2) );
											}
										}, true);
									}
									AsyncForEach(TblStruc.feildList, 0);
								}
							});
						} 
					})

					function Add2DOC (item, subOBJ, results2, callback, Last){
						var temp = item.subFeilds.feildList;
						if (! subOBJ[item.feildName]) subOBJ[item.feildName]=[];
						AsyncForEach2 = function(Items2, Index2) { //Result rows Iteration

							function checkAsyncForEach2Loop(callback, subOBJ) { //Call AsyncForEach2 loop till the end of Result rows
								//console.log("LoopStack", LoopStack, "subOBJ : ", subOBJ, callback)
								if (LoopStack.length>0) {
									LoopStack.some(function (x, i) { 
										topLevel=LoopStack.length-1;
										PreviousLoop_OBJ=LoopStack[topLevel];
										if (PreviousLoop_OBJ.Items.length-1>PreviousLoop_OBJ.Index) {	
											AsyncForEach2(PreviousLoop_OBJ.Items, ++LoopStack[i].Index); 
											return true;
										} else {
											LoopStack.pop();
											if (LoopStack.length==0) {
												callback(subOBJ);	//End of loop return !!!
											}
										}
									});	
								} else {
									callback(subOBJ); //End of loop return !!!
								}			
							}	

							result=Items2[Index2]; 
						 	subOBJ[item.feildName].push(result[item.feildName]); 

							if(temp instanceof Array) { 
								MakeQuery(item, result[item.feildName], function(err, Statement){
									if (!err) {
										ExecuteQuery(Statement, [], function (err3, results3) {
											if (err3 || results3.rows.length==0) {
												checkAsyncForEach2Loop(callback, subOBJ);
											} else {
												LoopStack.push({"Items" : Items2, "Index" : Index2, "feildName" : item.feildName})
												subOBJ[item.feildName].push({});
												topLevel=subOBJ[item.feildName].length-1;
												var a=subOBJ[item.feildName][topLevel];
												AsyncForEach3 = function(Items3, Index3) {
													Add2DOC(Items3[Index3], a, results3.rows, function () {
														if (Items3.length-1>Index3) {
															AsyncForEach3(Items3, ++Index3);
														} else {
															checkAsyncForEach2Loop(callback, subOBJ);
														}	
													}, Items3.length-1>=Index3);
												}	
												AsyncForEach3(temp, 0);
											}
										});
									} else {
										checkAsyncForEach2Loop(callback, subOBJ);
									}
								});		
							} else {
								checkAsyncForEach2Loop(callback, subOBJ);
							}	 
						}
						AsyncForEach2(results2,0);
					}  
				}

				function MakeQuery(root, pk ,callback) {
					if ( pk ) {//
						var coma=Statement=""
						if (root.feildList) var feildList=root.feildList, table=root.table, primaryKey=root.primaryKey
						else var feildList=root.subFeilds.feildList, table=root.subFeilds.table, primaryKey=root.subFeilds.primaryKey;

						feildList.some(function(champ, index){
					        Statement=Statement+coma+champ.feildName+" "+champ.feildName;
					        coma=", ";
					        if (index==feildList.length-1) { 
					        	Statement="select "+primaryKey+" PRIMARYKEY, "+Statement+" FROM "+table+" WHERE "+primaryKey+" = '"+pk+"'";
								callback(false, Statement);
						    }        	
						});	
					} else {
						callback(true, pk); //Error when primaryKey value is null
					}
				}


				function getConnexion(callBack) {
					config.Oracle_POOL.getConnection( function(err, connection) {
						if (err){ 
							console.log("getConnection err --> ", err);
							if (connection)
							connection.release(
				                function(err2) {
				                	if (err2) {
				                		console.log("getConnection -->  connection.release  --> err",err)
				                	}
				                }
				            );  
				            if (typeof callBack=="function") callBack(err, null);  	
						} else {

						 	// NUMBER OF CONNCETIONS OPEN
						    console.log("AFTER/getConnexion -> ORACLE: CONNX OPEN: " + config.Oracle_POOL.connectionsOpen);

						    // NUMBER OF CONNEXTIONS IN USE
						    console.log("AFTER/getConnexion -> ORACLE: CONNX IN USE: " + config.Oracle_POOL.connectionsInUse);

						    console.log("poolMax", config.Oracle_POOL.poolMax,  oracle.poolMax)

							config.Connexions=connection;
							if (typeof callBack=="function") callBack(err, connection); 
						}	
						
					});	
				}	

				function releaseConnection (callBack) {
					//Committing pending statement before releasing...
					config.Connexions.execute("select 1 from dual", [], {autoCommit : true} , function(err, results) {
						if (config.stmtCounter) config.stmtCounter.endAt = new Date();
						console.log("AutoComit at releaseConnection ** config.stmtCounter : ", config.stmtCounter);
						
						config.Connexions.release(function(err2) {
				            	if (err2) {
				            		console.log("config.Connexions.release --> err",err2);
				            	} else console.log("Connexion released");
				            	if (typeof callBack=="function") callBack(err, "Commit failed");  
				        });  
					})	
				}	


				function ExecuteQuery(Query, LstParam, callBack) {
					if (config.Connexions){ 
						var OracleConfig={outFormat: oracle.OBJECT, maxRows : 200};
						
						if (!config.stmtCounter) config.stmtCounter={ "nonCommitable" : 0, Commitable : 0, startAt : new Date()};
						(/(^[ ]*Select)/gi).test(Query)?config.stmtCounter.nonCommitable++:config.stmtCounter.Commitable++;
						
						//Comming every 100 insert and update statement.
						if (config.stmtCounter.Commitable>100) {
							config.stmtCounter.Commitable=0;
							OracleConfig.autoCommit = true;
						}
						config.Connexions.execute(Query, LstParam, OracleConfig , function(err, results) {

					    	if (err && (/(ORA-01000)/gi).test(" "+err+" ")) 
					    		console.log("ORA-01000 error on stmtCounter : ", config.stmtCounter, " Query : ", Query);

							if (typeof callBack=="function") callBack(err, results);
							//setTimeout(function() {
							//	exit(0);
							//}, 60000)
						});	
					} else {
						if (typeof callBack=="function") callBack(true, results, "no connexion in config.Connexions "); 	
					}

				}

				function doCreatePool (callback) {
				    dbConfig=config.Settings.DataBaseConnexionString;
				    oracle.createPool (
				      {
				        user          : dbConfig.user,
				        password      : dbConfig.password,
				        connectString : dbConfig.connectString,
				        poolMax       : 44,
				        poolMin       : 2,
				        poolIncrement : 5,
				        poolTimeout   : 2,
				        stmtCacheSize : 200
				      },
				      function(err, pool)
				      {
				        if (err) {
				          console.error("createPool() callback: " + err.message);
				          return;
				        } else {
				            config.Oracle_POOL = pool;
				        }
				        if (typeof callback=="function") callback(err, pool);
				     });      
				}

**Comment by dmcghan** [comment]
@BlackMatrix971 Your demo code only calls getConnexion once and then uses setTimeout to simulate some even happening multiple times. But wouldn't that have to keep reusing the same connection? Earlier you said, "When an other event occur I call getConnexion."

Also, before I forget, I don't know too much about what's going on here, but look out for SQL injection vulnerabilities. The type of dynamic statements you are generating a susceptible to them. You should create some assertion routines that make sure the values are what you expect them to be (valid object names) and throw errors if they are not.

**Comment by dmcghan** [comment]
@BlackMatrix971 Some other notes...

* You have a function named releaseConnection that is never used. 
* In releaseConnection you execute a query on dual with autoCommit set to true. You should just call the commit method of the connection.
* The name config.Connexions seems to indicate that there are multiple connections, but there's just one. I would rename that to avoid confusion.

**Comment by BlackMatrix971** [comment]
Yes it's true. the code should look like this : 

/*
=========================
==		APP Code		=
=========================
*/
console.log("initializing...")
doCreatePool (function(){
	getConnexion(function(){
		console.log("doCreatePool -> getConnexion");

		setInterval(function(){
		 	MakeDOC({PRIMARYKEY : "A0"}, TblStruc, function(err, data) {
		 		if (err) {
		 			console.log("MakeDOC -> err", err, data)
		 		}
		 		releaseConnection (function (err, msg) {

		 		});
		 	}); 
			 	//Simulate event :
			 	//	socket.on('MakeDOC', function(Response){		
				//		MakeDOC(Response, TblStruc...);
				//	})
		}, 3000);

	})
})

**Comment by BlackMatrix971** [comment]
And connexionRelease becomes :

function releaseConnection (callBack) {
	//Committing pending statement before releasing...
	config.Connexions.commit(function(err) {
		if (config.stmtCounter) config.stmtCounter.endAt = new Date();
		console.log("AutoComit at releaseConnection ** config.stmtCounter : ", config.stmtCounter);
		
		config.Connexions.release(function(err2) {
            	if (err2) {
            		console.log("config.Connexions.release --> err",err2);
            	} else console.log("Connexion released");
            	if (typeof callBack=="function") callBack(err, "Commit failed");  
        });  
	})	
}

**Comment by BlackMatrix971** [comment]
@dmcghan 

    Also, before I forget, I don't know too much about what's going on here, but look out for SQL injection     vulnerabilities. The type of dynamic statements you are generating a susceptible to them. You should     create some assertion routines that make sure the values are what you expect them to be (valid object     names) and throw errors if they are not.


Ok thanks, I'll check that.

**Comment by BlackMatrix971** [comment]
@dmcghan 
    But wouldn't that have to keep reusing the same connection? 

I tried that already, but after a certain number of statement are executed, ORA-01000 falls.

**Comment by alexsantos** [comment]
ORA-01000 arises because you are not closing the statements, that's not
related to the connection. While testing the code, run this SQL on your
database, you'll see the number of cursors rising till you get ORA-01000:

--total cursors open, by session
select a.value, s.username, s.sid, s.serial#
from v$sesstat a, v$statname b, v$session s
where a.statistic# = b.statistic#  and s.sid=a.sid
and b.name = 'opened cursors current';


http://www.orafaq.com/node/758


2015-06-08 21:56 GMT+01:00 BlackMatrix971 <notifications@github.com>:

**Comment by BlackMatrix971** [comment]
Thanks I check that, How can I close the statements from nodejs ?

**Comment by BlackMatrix971** [comment]
I read that among the things I read concerning statement : 

Many of node-oracledb‘s features are immediately available to an application. For example, statement caching is enabled by default. The feature helps reduce unnecessary processing and network overhead of statement parsing when a statement is re-executed. Instead of requiring node-oracledb to have explicit parse and execute methods, by taking advantage of Oracle’s client library statement caching, the node-oracledb API is simplified.

But I don't find how to close the statements this would be the solution of my problem.

**Comment by cjbj** [comment]
@alexsantos @BlackMatrix971 in node-oracledb 0.6 which fetches all data from a query in one shot (as opposed to the result set model that is planned), the driver closes the statement for you.

Can you monitor pool usage and see how it is growing?  And count the number of distinct SQL statements. There will be a statement cache per session.  Your 44 sessions x 200 statements could be larger than max_cursors value that the DB is configured with.

I'm at an offsite this week so haven't been able to look at the code.

**Comment by alexsantos** [comment]
I see. My main experience with Oracle is with PLSQL and Java so I'm a bit
biased :-)
 Em 08/06/2015 22:35, "Christopher Jones" <notifications@github.com>
escreveu:

**Comment by BlackMatrix971** [comment]
@cjbj 

Hes I did it already :

the variable config.stmtCounter is used for that:

By adding different peaces of node logs, about 56000 select statements and 2000 insert statement are executed before I  reach ORA-24418 ** ORA-24418 **  ORA-01000 errors :

AutoComit at releaseConnection ** config.stmtCounter[conXKey] :  { nonCommitable: 2258,
  Commitable: 99,
  startAt: Fri Jun 05 2015 08:22:47 GMT-0400 (AST),
  endAt: Fri Jun 05 2015 08:23:13 GMT-0400 (AST) } 
Connexion released

erreur de connexion Oracle -->  [Error: ORA-24418: Cannot open further sessions.]
getConnexion failed ! err :  [Error: ORA-24418: Cannot open further sessions.]
-- ORA-01000 error on stmtCounter :  3659

**Comment by cjbj** [comment]
@BlackMatrix971 Without digging into code I have two theories: 
1) you are not releasing sessions back to the pool
2) your session pool is too small for the throughput Node is capable of.
Either way, your statement cache size x connection pool size is likely > the DB open_cursors parameter.  This needs to be fixed by adjusting some or all of those three settings.
Of course, there could be a bug in node-oracledb which is why the testcase is useful.

**Comment by cjbj** [comment]
@BlackMatrix971 can you let us know whether you are also using DRCP?

**Comment by dmcghan** [comment]
@BlackMatrix971 After making the changes you suggested I also changed the timeout to 1 second. The code ran for over 16 hours without any issues. The number of open connections never exceeded 2 and the connections in use never exceeded 1. I am unable to reproduce the error with this test.

However, @cjbj's theory about the statement cache size makes a lot of sense because your code could generate many dynamic queries (something the test doesn't show). Why not try setting the stmtCacheSize to 0 to see if you still get the ORA-01000?

**Comment by BlackMatrix971** [comment]
@cjbj 

I'm not a specialist in data base, I just use pool connexion as the exemple stated for my own need.
I don't know about DRCP.
I read a lot those day to update my self. I spent more time reading than programming ;-)
I just use simple pool connexion I think.

**Comment by BlackMatrix971** [comment]
This guy face the same problem...

http://codegur.com/29846188/node-js-express-oracle-connection-pooling-ora-24418-cannot-open-further-sess

**Comment by dmcghan** [comment]
@BlackMatrix971 Were you able to test again with the stmtCacheSize set to 0?

**Comment by cjbj** [comment]
@BlackMatrix971 or test with a smaller pool size and smaller cache size.  And optionally with a bigger open_cursors value in the DB.

**Comment by BlackMatrix971** [comment]
It's ok.
I founded a solution.
I'll post it when I'm back.

Thanks to all for you support.

**Comment by mrchess** [comment]
@BlackMatrix971 What was your solution? i am also running into `Error: ORA-01000: maximum open cursors exceeded]` errors despite releasing my connection after each query and having a cursor count of `1000`.

**Comment by cjbj** [comment]
@mrchess you might want to open a new issue and show us a reproducible testcase.
The statement cache size and your pool size are factors.
Don't forget to bump the DB parameter to allow enough cursors per session.

**Comment by krismohan** [comment]
If you are using resultsets, make sure you close them too, before releasing the connection to the pool. See https://github.com/oracle/node-oracledb/blob/master/doc/api.md#close

**Comment by dmcghan** [comment]
@mrchess Are you just executing queries or PL/SQL code too? Do your queries make calls to PL/SQL code? 

```sql
select my_custom_func(id)
from emps; 
```

You'll need to ensure that anything cursors opened by any PL/SQL calls are closed as well.

**Comment by cjbj** [comment]
With help from @krismohan here is a blog post on avoiding ORA-1000: https://blogs.oracle.com/opal/entry/node_oracledb_avoiding_ora_01000

**Comment by michael1389** [comment]
what was the solution to reduce open_cursors?
